File: DSA/src/Questions/Graph/ParallelCourses.java
package Questions.Graph;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

/*
https://leetcode.com/problems/parallel-courses/description/
 You are given an integer n, which indicates that there are n courses labeled from 1 to n. You are also given an array relations where relations[i] = [prevCoursei, nextCoursei], representing a prerequisite relationship between course prevCoursei and course nextCoursei: course prevCoursei has to be taken before course nextCoursei.

In one semester, you can take any number of courses as long as you have taken all the prerequisites in the previous semester for the courses you are taking.

Return the minimum number of semesters needed to take all courses. If there is no way to take all the courses, return -1.

 

Example 1:


Input: n = 3, relations = [[1,3],[2,3]]
Output: 2
Explanation: The figure above represents the given graph.
In the first semester, you can take courses 1 and 2.
In the second semester, you can take course 3.
Example 2:


Input: n = 3, relations = [[1,2],[2,3],[3,1]]
Output: -1
Explanation: No course can be studied because they are prerequisites of each other.
 

Constraints:

1 <= n <= 5000
1 <= relations.length <= 5000
relations[i].length == 2
1 <= prevCoursei, nextCoursei <= n
prevCoursei != nextCoursei
All the pairs [prevCoursei, nextCoursei] are unique.
 */

//  Time complexity: O(V+E)
//  Space complexity: O(V+E)
public class ParallelCourses {
    public int minimumSemesters(int n, int[][] relations) {
        ArrayList<Integer>[] adj = new ArrayList[n + 1];
        int[] indegree = new int[n + 1];

        for (int[] rel : relations) {
            indegree[rel[1]]++;
            if (adj[rel[0]] == null) {
                adj[rel[0]] = new ArrayList();
            }
            adj[rel[0]].add(rel[1]);
        }

        Queue<Integer> queue = new LinkedList<>();
        for (int i = 1; i <= n; i++)
            if (indegree[i] == 0)
                queue.offer(i);

        if (queue.isEmpty())
            return -1;

        int step = 0, visitedCount = 0;

        while (!queue.isEmpty()) {
            int size = queue.size();
            step++;
            while (size-- > 0) {
                visitedCount++;
                Integer cur = queue.poll();
                if (adj[cur] != null) {
                    for (int k : adj[cur]) {
                        indegree[k]--;
                        if (indegree[k] == 0)
                            queue.offer(k);
                    }
                }

            }
        }

        return visitedCount == n ? step : -1;
    }

    public static void main(String[] args) {
        int n = 3;
        int[][] relations = { { 1, 3 }, { 2, 3 } };
        ParallelCourses parallelCourses = new ParallelCourses();
        int result = parallelCourses.minimumSemesters(n, relations);
        System.out.println("Minimum Semesters: " + result);
    }

}

File: DSA/Python/Questions/Graph/TopologicalSort/ParallelCourses.py
"""
# Question: Parallel Courses
# Link: https://leetcode.com/problems/parallel-courses/

# Time Complexity: O(V + E)
# Space Complexity: O(V + E)

# Algorithm:
# 1. Build graph from course dependencies
# 2. Use BFS for level-wise traversal
# 3. Return minimum semesters needed
"""

from collections import defaultdict, deque


class ParallelCourses:
    def minimum_semesters(self, n: int, relations: list[list[int]]) -> int:
        # Build graph and calculate in-degree
        graph = defaultdict(list)
        in_degree = [0] * (n + 1)

        for prev, next_course in relations:
            graph[prev].append(next_course)
            in_degree[next_course] += 1

        # Start with courses having no prerequisites
        queue = deque([i for i in range(1, n + 1) if in_degree[i] == 0])
        semesters = 0
        courses_taken = 0

        while queue:
            size = len(queue)
            semesters += 1

            for _ in range(size):
                course = queue.popleft()
                courses_taken += 1

                for next_course in graph[course]:
                    in_degree[next_course] -= 1
                    if in_degree[next_course] == 0:
                        queue.append(next_course)

        return semesters if courses_taken == n else -1


def main():
    solution = ParallelCourses()
    relations = [[1, 2], [2, 3], [3, 1]]
    print(solution.minimum_semesters(3, relations))


if __name__ == "__main__":
    main()
